cli/cmd/cluster/cluster.go

```diff
@@ -341,3 +365,103 @@ func (c controlplaneUpdater) ensureComponent(component, namespace string) error
 
 	return nil
 }
+
+// taintCertificates taints all certificate resources in existing Terraform state.
+// it will not taint the private keys of the CA so the public key gets reused and the old CA cert can trust the new certificates.
+func (c *cluster) taintCertificates() error {
+	steps := []terraform.ExecutionStep{}
+
+	for _, t := range c.certificateResources() {
+		steps = append(steps, terraform.ExecutionStep{
+			Args: []string{"taint", t},
+		})
+	}
+
+	if err := c.terraformExecutor.Execute(steps...); err != nil {
+		return fmt.Errorf("tainting existing certificates: %w", err)
+	}
+
+	return nil
+}
+
+// untaintCertificates untaints all certificate resources in existing Terraform state.
+// it will be used to rollback taints on errors of taints on best effort basis.
+func (c *cluster) untaintCertificates() error {
+	steps := []terraform.ExecutionStep{}
+
+	for _, t := range c.certificateResources() {
+		steps = append(steps, terraform.ExecutionStep{
+			Args: []string{"untaint", t},
+		})
+	}
+
+	if err := c.terraformExecutor.Execute(steps...); err != nil {
+		return fmt.Errorf("untaint existing certificates: %w", err)
+	}
+
+	return nil
+}
+
+func (c *cluster) certificateResources() []string {
+	f := func(resourceName string) string {
+		m := c.platform.Meta()
+
+		return fmt.Sprintf("module.%s-%s.module.bootkube.%s", m.Name, m.ClusterName, resourceName)
+	}
+
+	targets := []string{
+		// certificates
+		"tls_locally_signed_cert.admin",
+		"tls_locally_signed_cert.admission-webhook-server",
+		"tls_locally_signed_cert.aggregation-client[0]",
+		"tls_locally_signed_cert.apiserver",
+		"tls_locally_signed_cert.client",
+		"tls_locally_signed_cert.kubelet",
+		"tls_locally_signed_cert.peer",
+		"tls_locally_signed_cert.server",
+		"tls_self_signed_cert.aggregation-ca[0]",
+		"tls_self_signed_cert.etcd-ca",
+		"tls_self_signed_cert.kube-ca",
+		// Taint non-CA private keys as well, as those are safe to rotate.
+		"tls_private_key.admin",
+		"tls_private_key.admission-webhook-server",
+		"tls_private_key.aggregation-client[0]",
+		"tls_private_key.apiserver",
+		"tls_private_key.client",
+		"tls_private_key.kubelet",
+		"tls_private_key.peer",
+		"tls_private_key.server",
+	}
+
+	var fullTargets []string
+	for _, target := range targets {
+		fullTargets = append(fullTargets, f(target))
+	}
```

Comment:
Wouldn't this be simpler than adding a function variable. Also avoid unnecessary function call to `c.platform.Meta()` on every iteration.

```suggestion
	m := c.platform.Meta()
	var fullTargets []string
	for _, target := range targets {
		fullTargets = append(fullTargets, fmt.Sprintf("module.%s-%s.module.bootkube.%s", m.Name, m.ClusterName, target))
	}
```

Reply:
Yes, I don't think the function variable is better nor makes things more clear.

Reply:
Clearly my brain fart, nicely spotted :smile: